library/Network/CloudSeeder/CommandLine.hs

```diff
@@ -1,28 +1,77 @@
 module Network.CloudSeeder.CommandLine
     ( Command(..)
-    , parseCommand
+    , ParameterSpec(..)
+    , parseArguments
+    , parseOptions
     , withInfo
     ) where
 
-import Options.Applicative (ArgumentFields, Parser, ParserInfo, Mod, subparser, strArgument, command, info, progDesc, metavar, helper)
+import Data.Monoid ((<>))
+import Options.Applicative 
+
+import qualified Data.Map as M
+import qualified Data.Set as S
 import qualified Data.Text as T
 
-data Command = DeployStack T.Text T.Text
+import Network.CloudSeeder.Types
+
+data Command = DeployStack 
+ { commandStack :: T.Text
+ , commandEnv :: T.Text
+ } deriving (Eq, Show)
+
+-- helpers --
 
 textArgument :: Mod ArgumentFields String -> Parser T.Text
 textArgument = fmap T.pack . strArgument
 
+textOption :: Mod OptionFields String -> Parser T.Text
+textOption = fmap T.pack . strOption
+
 withInfo :: Parser a -> String -> ParserInfo a
-withInfo opts desc = info (helper <*> opts) $ progDesc desc
+withInfo parser desc = info (helper <*> parser) $ progDesc desc
 
-parseDeploy :: Parser Command
-parseDeploy = DeployStack <$> parseStack <*> parseEnv
+-- parsers --
+
+parseOptions :: S.Set (T.Text, ParameterSource) -> ParserInfo (M.Map T.Text T.Text)
+parseOptions ps = info (helper <*> parseCommand *> parseParameters ps)
+  ( fullDesc 
+ <> progDesc "Interact with the CloudFormation API o"
+ <> header "Cloud-Seeder -- a tool for interacting with the AWS CloudFormation API"
+  )
+
+parseParameters :: S.Set (T.Text, ParameterSource) -> Parser (M.Map T.Text T.Text)
+parseParameters ps = M.fromList <$> traverse parseParameter (S.toList ps)
+
+parseParameter :: (T.Text, ParameterSource) -> Parser (T.Text, T.Text)
+parseParameter (key, Flag) = do 
+  let k = T.unpack key
+  val <- textOption (long k <> metavar k <> noArgError (ErrorMsg $ "Required flag not provided: " <> k))
+  pure (key, val)
+parseParameter (key, _) = do 
+  let k = T.unpack key
+      err = ErrorMsg "parseParameter -- the impossible happened: attempted to parse a non-flag parameter source."
+  abort <- abortOption err (long k <> metavar k)
+  pure (key, abort key)
+
+parseArguments :: ParserInfo Command
+parseArguments = info (helper <*> parseCommand) 
+  ( fullDesc 
+ <> progDesc "Interact with the CloudFormation API a"
+ <> header "Cloud-Seeder -- a tool for interacting with the AWS CloudFormation API"
+  )
 
 parseCommand :: Parser Command
 parseCommand = subparser $ command "deploy" (parseDeploy `withInfo` "Deploy a stack to ENV")
 
+parseDeploy :: Parser Command
+parseDeploy = DeployStack <$> parseStack <*> parseEnv <* many parseAnyOption
```

Comment:
Is `parseAnyOption` here necessary? Given that `getArgs` does the argument parsing on its own now.